### ISSUE TITLE
parametrization of non-convex polygons, flat disks and flat spheres

### ISSUE THREAD

**Issue by videlec** [videlec/sage-flatsurf#31]
The `EquiangularPolygons` class allows to play with polygons where angles are fixed. However, it does not handle gracefully the non-convex case where intersections might occur. The parameter space for the lengths is an a priori complicated space since each intersection of segments gives a quadratic equation.

Perhaps, non-convex polygons is the wrong question and one should look at equiangular flat disks. Also, most of the times, these polygons and flat disks will be used in their "doubled" version which is a sphere with conical flat metrics.

In these three settings (polygons with fixed boundary angles, flat disks with fixed boundary angles, flat spheres with fixed conical angles) it would be nice to implement algorithms for
- producing an example
- producing many examples (possibly involving randomness)
- parametrizing the space (e.g. provide a reasonable cell decomposition in which one can "walk")

**Comment by videlec** [comment]
@wphooper Pat, do you know the answer to the "possibly interesting question" ? I think it is worth investigating how much the boundary gives you information on a flat disk.

**Comment by videlec** [comment]
There is a numerical approach to this. Consider a tuple of integers `(a_1, a_2, ..., a_k)` and let `N = a_1 + a_2 + ... + a_k`. We consider the associated angles `(k-2) pi / N (a_1, a_2, ..., a_k)`. Then a flat disk with these angles is nothing more than a (multivalued) differential
```
(z - p_1)^(a_1 - N/k) (z - p_2)^(a_2 - N/k) ... (z - p_k)^(a_k - N/k) dz
```
with `p1`, ..., `pk` reals (and one can fix `p_1=0`, `p_2=1`, `p_3=oo` to have a bijection with flat disks with fixed angles). To obtain the corresponding lengths one has to integrate the differential between consecutive points `p_i` on the circle. The absolute value of the integral is well defined whatever choice of branch is made.

One can even compute a combinatorial triangulation (close to the Delaunay cell decomposition) by integrating the linear flow.

If there are two flat disks with the same boundary it means that there are two tuples of `p_i` with the same ordering on the circle for which all integrals coincide (up to a global multiplicative scalar which corresponds to the area) !

**Comment by saraedum** [comment]
This is related to https://github.com/flatsurf/sage-flatsurf/pull/170.

I'm not sure what you had in mind here @videlec. Should flat disks (and spheres?) be fundamental building blocks of surfaces like polygons are currently? Or do you just want to create a flat disk with a convenience function that produces a gadget that is built from polygons glued to the flat disk and that can then be used to build more complex surfaces.

**Comment by videlec** [comment]
I don't think that flat disks should be any special : they are translation surfaces with boundary (that happen to be homeomorphic to a disk). I just wanted some convenience functions. In particular, given boundary segments, it should be possible to build all flat disks whose boundary is made of the segments.

**Comment by saraedum** [comment]
I don't understand this comment.

> There is a numerical approach to this. Consider a tuple of integers `(a_1, a_2, ..., a_k)` and let `N = a_1 + a_2 + ... + a_k`. We consider the associated angles `(k-2) pi / N (a_1, a_2, ..., a_k)`. Then a flat disk with these angles is nothing more than a (multivalued) differential
> 
> ```
> (z - p_1)^(a_1 - N/k) (z - p_2)^(a_2 - N/k) ... (z - p_k)^(a_k - N/k) dz
> ```
> 
> with `p1`, ..., `pk` reals

So, if I take a₁, a₂, a₃ = 1 (an equilateral triangle), I just get dz. From what you wrote it would seem that I can choose the p<sub>k</sub> however I want but then I won't get three equal side lengths for my equilateral triangle.

> (and one can fix `p_1=0`, `p_2=1`, `p_3=oo` to have a bijection with flat disks with fixed angles).

I don't understand this. Didn't you say above that I get a flat disk with these angles already without fixing the points like that?

Sorry, for asking these questions two years after the fact but I was looking into whether it's worthwhile to implement this now and I would like to understand the idea here.
